wasmtime-wasi/src/host_impls.rs

```diff
@@ -92,6 +92,33 @@ pub fn wasmtime_ssp_proc_exit(rval: wasm32::__wasi_exitcode_t) {
     ::std::process::exit(rval as i32)
 }
 
+pub unsafe fn wasmtime_ssp_fd_prestat_get(
+    prestats: *mut host::fd_prestats,
+    fd: host::__wasi_fd_t,
+    buf: *mut host::__wasi_prestat_t,
+) -> host::__wasi_errno_t {
+    host::rwlock_rdlock(&mut (*prestats).lock as *mut host::rwlock);
+
+    let mut prestat: *mut host::fd_prestat = ::std::ptr::null_mut();
+    let error = host::fd_prestats_get_entry(prestats, fd, &mut prestat);
+
+    if error != host::__WASI_ESUCCESS {
+        host::rwlock_unlock(&mut (*prestats).lock as *mut host::rwlock);
+        return error;
+    }
+
+    (*buf).pr_type = host::__WASI_PREOPENTYPE_DIR as host::__wasi_preopentype_t;
+    let dir_name = ::std::ffi::CStr::from_ptr((*prestat).dir).to_str();
+    if let Err(_) = dir_name {
+        host::rwlock_unlock(&mut (*prestats).lock as *mut host::rwlock);
+        return host::__WASI_EBADF;
```

Comment:
Here, I have to admit that I have absolutely no clue which error code would be best to return, or if it is even reasonable to convert raw C ptr to `&str` in order to get its length. We could use `libc::strlen` but to me that sounds even worse.

Reply:
These prestat directory names are things we can validate at wasmtime startup, so this code can actually just panic if it somehow sees an invalid string here.

Concerning the CStr, one thing we could do instead here is replace the `dir` member of prestat with a pointer and a length so that we can compute it once up front rather than at runtime.

Reply:
That sounds good. I'll give it a spin then!

Reply:
OK, so I've modified `fd_prestat` to hold C-str together with its length, and both are populated once in `fd_prestats_insert`. Have a look and lemme know what you reckon!